dss/api/collections.py

```diff
@@ -0,0 +1,185 @@
+        raise DSSException(404, "not_found", "Could not find collection for UUID {}".format(uuid))
+    return json.loads(collection_blob)
+
+@dss_handler
+def get(uuid: str, replica: str, version: str = None):
+    authenticated_user_email = request.token_info['email']
+    collection_body = get_impl(uuid=uuid, replica=replica, version=version)
+    if collection_body["owner"] != authenticated_user_email:
+        raise DSSException(requests.codes.forbidden, "forbidden", f"Collection access denied")
+    return collection_body
+
+@dss_handler
+def find(replica: str):
+    raise NotImplementedError()
+
+@dss_handler
+def put(json_request_body: dict, replica: str, uuid: str, version: str):
+    authenticated_user_email = request.token_info["email"]
+    collection_body = dict(json_request_body, owner=authenticated_user_email)
+    uuid = uuid.lower()
+    handle = Config.get_blobstore_handle(Replica[replica])
+    verify_collection(collection_body["contents"], Replica[replica], handle)
+    collection_uuid = uuid if uuid else str(uuid4())
+    if version is not None:
+        # convert it to date-time so we can format exactly as the system requires (with microsecond precision)
+        timestamp = iso8601.parse_date(version)
+    else:
+        timestamp = datetime.datetime.utcnow()
+    collection_version = datetime_to_version_format(timestamp)
+    handle.upload_file_handle(Replica[replica].bucket,
+                              "{}/{}.{}".format(COLLECTION_PREFIX, collection_uuid, collection_version),
+                              io.BytesIO(json.dumps(collection_body).encode("utf-8")))
+    return jsonify(dict(uuid=collection_uuid, version=collection_version)), requests.codes.created
+
+class hashabledict(dict):
+    def __hash__(self):
+        return hash(tuple(sorted(self.items())))
+
+@dss_handler
+def patch(uuid: str, json_request_body: dict, replica: str, version: str):
+    # patch one collection with another
+    uuid = uuid.lower()
+    handle = Config.get_blobstore_handle(Replica[replica])
+    try:
+        cur_collection_blob = handle.get(Replica[replica].bucket, "{}/{}.{}".format(COLLECTION_PREFIX, uuid, version))
+    except BlobNotFoundError:
+        raise DSSException(404, "not_found", "Could not find collection for UUID {}".format(uuid))
+    collection = json.loads(cur_collection_blob)
+    for field in "name", "description", "details":
+        if field in json_request_body:
+            collection[field] = json_request_body[field]
+    remove_contents_set = set(map(hashabledict, json_request_body.get("removeContents", [])))
+    collection["contents"] = [i for i in collection["contents"] if hashabledict(i) not in remove_contents_set]
+    verify_collection(json_request_body.get("addContents", []), Replica[replica], handle)
+    collection["contents"].extend(json_request_body.get("addContents", []))
+    timestamp = datetime.datetime.utcnow()
+    new_collection_version = datetime_to_version_format(timestamp)
```

Comment:
Not idempotent. See @ttung's and my comments above in the Swagger def.

Reply:
It's not idempotent in the sense that it creates an orphan object, but as we discussed offline - I *think* that's not a big deal and orphaned collection objects should not matter. If there's a use case where this becomes an obvious issue, I'd be happy to reconsider. But as per the current design, you have to retrieve the newly assigned version to do anything with the result, which I hope means the API user is naturally guided toward the right pattern. I will look for places to add more docs to clarify this (added another checkbox in #1269).